### PR TITLE
SDK-3824 Refactor adapter to start using profile attribute names

### PR DESCRIPTION
**Pull request by kushCT**
- Refactor Event adapter and Trigger Adapter to start using profile attribute names
- Added support for User Attribute triggering for Profile Increment and Profile Decrement API

### CODE REVIEW

File: CleverTapSDK/CTUserInfoMigrator.m [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    // Construct the old plist file name and path
+    NSString *userProfileWithDeviceID = [NSString stringWithFormat:@"clevertap-%@-userprofile.plist", acc_id];
+    NSString *userProfileWithDeviceIDPath = [libraryPath stringByAppendingPathComponent:userProfileWithDeviceID];
+        
+    // Construct the new plist file name and path
+    NSString *userProfileWithDeviceIDAndGUID = [NSString stringWithFormat:@"clevertap-%@-%@-userprofile.plist", acc_id, device_id];
+    NSString *userProfileWithDeviceIDAndGUIDPath = [libraryPath stringByAppendingPathComponent:userProfileWithDeviceIDAndGUID];
+    
+    NSError *error = nil;
+    if ([fileManager fileExistsAtPath:userProfileWithDeviceIDAndGUIDPath]) {
```
Comment (nzagorchev) [inline CleverTapSDK/CTUserInfoMigrator.m:27]: Check if old file exists, migrate and delete it. This will make one call to fileExists.

**Review by nzagorchev — COMMENTED** [review]
Please address comments and add unit tests.

File: CleverTapSDK/CTConstants.h [hunk 2]
```diff
@@ -241,6 +257,7 @@ extern NSString *CLTAP_PROFILE_IDENTITY_KEY;
 #pragma mark Constants for Profile identifier keys
 #define CLTAP_PROFILE_IDENTIFIER_KEYS @[@"Identity", @"Email"] // LEGACY KEYS
 #define CLTAP_ALL_PROFILE_IDENTIFIER_KEYS @[@"Identity", @"Email", @"Phone"]
+#define CLTAP_KeysToSkipForUserAttributesEvaluation @[@"cc", @"tz", @"Carrier"]
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTConstants.h:260]: Changes done

File: CleverTapSDK/CTLocalDataStore.m [hunk 3]
```diff
@@ -88,9 +92,13 @@ - (BOOL)inBackgroundQueue {
...
-    localProfileForSession = [NSMutableDictionary dictionary];
-    // this will remove the old profile from the file system
-    [self _persistLocalProfileAsyncWithCompletion:nil];
+    localProfileForSession = [NSMutableDictionary new];
+    [self runOnBackgroundQueue:^{
+        @synchronized (self->localProfileForSession) {
+            self->localProfileForSession = [self _inflateLocalProfile];
+        }
+    }];
+//    [self _persistLocalProfileAsyncWithCompletion:nil];
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTLocalDataStore.m:101]: Changes done

File: CleverTapSDK/CleverTap.m [hunk 4]
```diff
@@ -2923,6 +2924,72 @@ - (void)_handleMultiValueProfilePush:(NSDictionary*)customFields updatedMultiVal
     }
 }
 
+- (NSDictionary<NSString *, NSDictionary<NSString *, id> *> *)getUserAttributeChangeProperties:(NSDictionary *)event {
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CleverTap.m:2927]: Changes done

File: CleverTapSDK/InApps/CTInAppEvaluationManager.m [hunk 5]
```diff
@@ -87,19 +87,39 @@ - (void)evaluateOnEvent:(NSString *)eventName withProps:(NSDictionary *)properti
...
+
+-(void)evaluateOnUserAttributeChange:(NSDictionary<NSString *, NSDictionary *> *)profile {
+    NSDictionary *appFields = self.appLaunchedProperties;
+    NSMutableArray<CTEventAdapter *> *eventAdapterList = [NSMutableArray array];
+    [profile enumerateKeysAndObjectsUsingBlock:^(NSString *key, id value, BOOL *stop) {
+        NSString *eventName = [key stringByAppendingString:CLTAP_USER_ATTRIBUTE_CHANGE];
+        CTEventAdapter *event = [[CTEventAdapter alloc] initWithEventName:eventName profileAttrName:key eventProperties: value andLocation:self.location];
+        NSMutableDictionary *mergedProperties = [event.eventProperties mutableCopy];
+        [mergedProperties addEntriesFromDictionary:appFields];
+        event.eventProperties = [mergedProperties copy];
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/CTInAppEvaluationManager.m:103]: Changes done

File: CleverTapSDK/InApps/Matchers/CTEventAdapter.m [hunk 6]
```diff
@@ -53,7 +53,7 @@ + (void)initialize {
 }
 
 - (instancetype)initWithEventName:(NSString *)eventName
-                  eventProperties:(NSDictionary *)eventProperties
+                  eventProperties:(NSMutableDictionary *)eventProperties
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/Matchers/CTEventAdapter.m:56]: Changes done

File: CleverTapSDK/InApps/Matchers/CTEventAdapter.m [hunk 7]
```diff
@@ -62,7 +62,7 @@ - (instancetype)initWithEventName:(NSString *)eventName
...
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/Matchers/CTEventAdapter.m:65]: Changes done

File: CleverTapSDK/CTUserInfoMigrator.m [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    // Get the path to the Documents directory
+    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
+    NSString *libraryPath = [paths objectAtIndex:0];
+    
+    // Construct the old plist file name and path
+    NSString *userProfileWithDeviceID = [NSString stringWithFormat:@"clevertap-%@-userprofile.plist", acc_id];
+    NSString *userProfileWithDeviceIDPath = [libraryPath stringByAppendingPathComponent:userProfileWithDeviceID];
+        
+    // Construct the new plist file name and path
+    NSString *userProfileWithDeviceIDAndGUID = [NSString stringWithFormat:@"clevertap-%@-%@-userprofile.plist", acc_id, device_id];
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTUserInfoMigrator.m]: Changes done

File: CleverTapSDK/CTUserInfoMigrator.m [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
++ (void)migrateUserInfoFileForAccountID:(NSString *)acc_id deviceID:(NSString *)device_id {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    
+    // Get the path to the Documents directory
+    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
+    NSString *libraryPath = [paths objectAtIndex:0];
+    
+    // Construct the old plist file name and path
+    NSString *userProfileWithDeviceID = [NSString stringWithFormat:@"clevertap-%@-userprofile.plist", acc_id];
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTUserInfoMigrator.m]: Changes done

File: CleverTapSDK/CTUserInfoMigrator.m [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+//  CTUserInfoMigrator.m
+//
+//  Created by Kushagra Mishra on 29/05/24.
+//
+
+#import "CTUserInfoMigrator.h"
+
+@implementation CTUserInfoMigrator
+
++ (void)migrateUserInfoFileForAccountID:(NSString *)acc_id deviceID:(NSString *)device_id {
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTUserInfoMigrator.m:6]: Changes done

[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTUserInfoMigrator.m:6]: Changes done

File: CleverTapSDK/InApps/Matchers/CTEventAdapter.m [hunk 11]
```diff
@@ -74,6 +74,20 @@ - (instancetype)initWithEventName:(NSString *)eventName
     return self;
 }
 
+- (instancetype)initWithEventName:(NSString *)eventName
+                  profileAttrName:(NSString *)profileAttrName
+                  eventProperties:(NSMutableDictionary *)eventProperties
+                      andLocation:(CLLocationCoordinate2D)location{
+    
+    if (self = [super init]) {
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/Matchers/CTEventAdapter.m:82]: Changes done

File: CleverTapSDK/InApps/Matchers/CTTriggersMatcher.m [hunk 12]
```diff
@@ -30,7 +30,7 @@ - (BOOL)matchEventWhenTriggers:(NSArray *)whenTriggers event:(CTEventAdapter *)e
 }
 
 - (BOOL)match:(CTTriggerAdapter *)trigger event:(CTEventAdapter *)event {
-    if (![[event eventName] isEqualToString:[trigger eventName]]) {
+    if (![[event eventName] isEqualToString:[trigger eventName]] && ([event profileAttrName] == nil || ![[event profileAttrName] isEqualToString:[trigger profileAttrName]])) {
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/Matchers/CTTriggersMatcher.m:33]: Changes done

File: CleverTapSDK/InApps/CTInAppEvaluationManager.m [hunk 13]
```diff
@@ -116,8 +136,15 @@ - (void)evaluateOnAppLaunchedServerSide:(NSArray *)appLaunchedNotifs {
...
 }
 
-- (void)evaluateClientSide:(CTEventAdapter *)event {
-    NSMutableArray *eligibleInApps = [self evaluate:event withInApps:self.inAppStore.clientSideInApps];
+- (void)evaluateClientSide:(NSArray<CTEventAdapter *> *)events {
+    NSMutableArray<NSDictionary *> *eligibleInApps = [NSMutableArray array];    
+    for (CTEventAdapter *event in events) {
+        id oldValue = [event.eventProperties objectForKey:CLTAP_KEY_OLD_VALUE];
+        id newValue = [event.eventProperties objectForKey:CLTAP_KEY_NEW_VALUE];
+        if (newValue == nil || newValue !=  oldValue){
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/InApps/CTInAppEvaluationManager.m:143]: Changes done

File: CleverTapSDK/CleverTap.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CleverTap.m:2927]: Changes done

File: CleverTapSDK/CTProfileBuilder.m [hunk 14]
```diff
@@ -453,5 +443,66 @@ + (void)_handleIncrementDecrementValue:(NSNumber *_Nonnull)value forKey:(NSStrin
     completion(operatorDict, newValue, nil);
 }
 
++ (NSNumber *_Nullable)_getUpdatedValue:(NSNumber *_Nonnull)value forKey:(NSString *_Nonnull)key withCommand:(NSString *_Nonnull)command cachedValue:(id)cachedValue {
+
+    NSNumber *newValue;
+    
+    if ([cachedValue isKindOfClass: [NSNumber class]]) {
```
[the remark this answers is not in the document]
Comment (kushCT) [reply CleverTapSDK/CTProfileBuilder.m:450]: Changes done

File: CleverTapSDK/CTUserInfoMigrator.m (hunk 1 above)
Comment (kushCT) [reply CleverTapSDK/CTUserInfoMigrator.m]: Changes done

**Review by nzagorchev — APPROVED** [review]
Please add unit tests before opening a PR to `develop`.

**Pull request merged**